src/Heap.sol

```diff
@@ -32,209 +36,171 @@ library BasicHeap {
     /// @param _value The value of the account to insert.
-    function insert(
-        Heap storage _heap,
-        address _id,
-        uint256 _value
-    ) internal {
-        if (_value == 0) revert WrongValue();
+    function insert(Heap storage _heap, address _id, uint256 _value) internal {
         if (_id == address(0)) revert AddressIsZero();
-        if (_heap.ranks[_id] != 0) revert AccountAlreadyInserted();
 
-        // Put the account at the end of the heap.
-        _heap.accounts.push(Account(_id, _value));
-        uint256 size = getSize(_heap);
-        _heap.ranks[_id] = size;
+        uint256 accountsLength = _heap.accounts.length;
+        if (_heap.indexOf[_id] != 0 || (accountsLength > 0 && _heap.accounts[0].id == _id)) {
+            revert AccountAlreadyInserted();
+        }
+
+        _heap.accounts.push();
 
-        // Restore the invariant.
-        shiftUp(_heap, size);
+        shiftUp(_heap, Account(_id, _value), accountsLength);
     }
 
     /// @notice Decreases the amount of an account in the `_heap`.
     /// @param _heap The heap to modify.
     /// @param _id The address of the account to decrease the amount.
     /// @param _newValue The new value of the account.
-    function decrease(
-        Heap storage _heap,
-        address _id,
-        uint256 _newValue
-    ) internal {
-        uint256 rank = _heap.ranks[_id];
-        if (rank == 0) revert AccountDoesNotExist();
-        uint256 oldValue = getAccount(_heap, rank).value;
-        if (_newValue >= oldValue || _newValue == 0) revert WrongValue();
-
-        setAccountValue(_heap, rank, _newValue);
-        shiftDown(_heap, rank);
+    function decrease(Heap storage _heap, address _id, uint256 _newValue) internal {
+        uint256 index = _heap.indexOf[_id];
+        if (index == 0 && (_heap.accounts.length == 0 || _heap.accounts[0].id != _id)) revert AccountDoesNotExist();
+
+        uint256 oldValue = _heap.accounts[index].value;
+        if (_newValue >= oldValue) revert WrongValue();
+
+        shiftDown(_heap, _heap.accounts.length, Account(_id, _newValue), index);
     }
 
     /// @notice Increases the amount of an account in the `_heap`.
     /// @dev Only call this function when `_id` is in the `_heap` with a smaller value than `_newValue`.
     /// @param _heap The heap to modify.
     /// @param _id The address of the account to increase the amount.
     /// @param _newValue The new value of the account.
-    function increase(
-        Heap storage _heap,
-        address _id,
-        uint256 _newValue
-    ) internal {
-        uint256 rank = _heap.ranks[_id];
-        if (rank == 0) revert AccountDoesNotExist();
-        uint256 oldValue = getAccount(_heap, rank).value;
+    function increase(Heap storage _heap, address _id, uint256 _newValue) internal {
+        uint256 index = _heap.indexOf[_id];
+        if (index == 0 && (_heap.accounts.length == 0 || _heap.accounts[0].id != _id)) revert AccountDoesNotExist();
```

Comment:
Maybe we should also store the length of accounts in this function?

Reply:
Since it's not used after, I don't think it's useful